exporter/src/main/as/flump/export/TexturePacker.as

```diff
@@ -20,314 +30,126 @@ public class TexturePacker
-    public const atlases :Vector.<Atlas> = new <Atlas>[];
-
-    public function PackerImpl (libs :Vector.<XflLibrary>, baseScale :Number, scaleFactor :int,
-        textureBorderSize :int, maxAtlasSize :int, optimizeForSpeed :Boolean,
-        quality :String, filenamePrefix :String) {
-
-        _textureBorderSize = textureBorderSize;
-        _maxAtlasSize = maxAtlasSize;
-        _optimizeForSpeed = optimizeForSpeed;
-        _quality = quality;
-
-        var scale :Number = baseScale * scaleFactor;
-
-        var useNamespaces :Boolean = libs.length > 1;
-        for each (var lib :XflLibrary in libs) {
+        var scale :Number = _baseScale * _scaleFactor;
+        var useNamespaces :Boolean = _libs.length > 1;
+        for each (var lib :XflLibrary in _libs) {
             for each (var tex :XflTexture in lib.textures) {
-                _unpacked.push(SwfTexture.fromTexture(lib, tex, quality, scale, useNamespaces));
+                _unpacked.push(SwfTexture.fromTexture(lib, tex, _quality, scale, useNamespaces));
             }
             for each (var movie :MovieMold in lib.movies) {
                 if (!movie.flipbook) continue;
                 for each (var kf :KeyframeMold in movie.layers[0].keyframes) {
-                    _unpacked.push(SwfTexture.fromFlipbook(lib, movie, kf.index, quality, scale,
-                        useNamespaces));
+                    _unpacked.push(SwfTexture.fromFlipbook(lib, movie, kf.index, _quality, scale,
+                            useNamespaces));
                 }
             }
         }
-        _unpacked.sort(Comparators.createReverse(Comparators.createFields(["a", "w", "h"])));
 
-        var start :int = getTimer();
-        while (_unpacked.length > 0) {
-            // Add a new atlas
-            const size :Point = findOptimalSize();
-            atlases.push(new AtlasImpl(
-                filenamePrefix + "atlas" + atlases.length,
-                size.x, size.y,
-                textureBorderSize,
-                scaleFactor,
-                quality));
-            var hasEmptyAtlas :Boolean = true;
-
-            // Try to pack each texture into any atlas
-            for (var ii :int = 0; ii < _unpacked.length; ++ii) {
-                var unpacked :SwfTexture = _unpacked[ii];
-                var w :int = unpacked.w + (_textureBorderSize * 2);
-                var h :int = unpacked.h + (_textureBorderSize * 2);
+        _unpacked.sort(Comparators.createReverse(Comparators.createFields(["a", "w", "h"])));
 
-                if (w > _maxAtlasSize || h > _maxAtlasSize) {
-                    throw new Error("Too large to fit in an atlas: '" + unpacked.symbol + "' (" +
-                        w + "x" + h + ")");
+        validateTextureSize(_maxAtlasSize);
+
+        var minAtlasSize : uint = calculateMinimumSize(0);
+        packIntoAtlas(minAtlasSize, 0);
+
+        return atlases;
+    }
+
+    private function packIntoAtlas(atlasSize : uint, currentPos : uint) : void {
```

Comment:
The new packing logic. It first tries to pack the textures into the smallest rectangle possible, if it fails it tries again with a twice at big rectangle. If everything still does not fit it creates  new texture for the stuff that did not fit and starts again